File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

string solution(vector<string> participant, vector<string> completion) {
    string answer = "";
    map<string,int> m;
    for(auto participant_person:participant){
        m[participant_person] += 1;
    }
    
    for(auto completion_person:completion){
        m[completion_person] -= 1;
    }
    
    for(auto participant_person:participant){
        if(m[participant_person]!=0){
            answer = participant_person;
            break;
        }
        
    }
    return answer;
}

File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.java
/*
풀이 방법
map을 이용해여 participant을 순회하면서 사람이름을 key 그리고 숫자를 1씩 높인다.
completion을 순회하여 사람이름을 key로 하여 숫자를 1씩 뺀다. 
숫자가 1이상이면 완주하지 못한 사람 이름을 반환한다.
*/

import java.util.HashMap;

class Solution {
    public String solution(String[] participant, String[] completion) {
        String answer = "";
        HashMap<String, Integer> personMap = new HashMap<>();
        
        for(String person : participant){
            personMap.put(person, personMap.getOrDefault(person, 0)+1);
        }
        
        for(String person : completion){
            personMap.put(person, personMap.getOrDefault(person, 0)-1);
        }
        
        for(String person : personMap.keySet()){
            if(personMap.get(person)!=0){
                return person;
            }
        }
        
        return answer;
    }
}